### PR TITLE
sdboot-builder: fix crash in exception handling

### PR DESCRIPTION
**Pull request by sternenseemann**
###### Motivation for this change

Extracted the fix from #156798, so it may be reviewed, merged and backported quicker.

cc @m-bdf 

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Review by m-bdf — APPROVED** [review]

File: nixos/modules/system/boot/loader/systemd-boot/systemd-boot-builder.py [hunk 1]
```diff
@@ -15,9 +15,12 @@
 import datetime
 import glob
 import os.path
-from typing import Tuple, List, Optional
+from typing import NamedTuple, List, Optional
 
-SystemIdentifier = Tuple[Optional[str], int, Optional[str]]
+class SystemIdentifier(NamedTuple):
```
Comment (lukegb) [inline nixos/modules/system/boot/loader/systemd-boot/systemd-boot-builder.py:20]: TIL. (no action required here)

File: nixos/modules/system/boot/loader/systemd-boot/systemd-boot-builder.py [hunk 2]
```diff
@@ -151,7 +154,7 @@ def get_generations(profile: Optional[str] = None) -> List[SystemIdentifier]:
     gen_lines.pop()
 
     configurationLimit = @configurationLimit@
-    configurations: List[SystemIdentifier] = [ (profile, int(line.split()[0]), None) for line in gen_lines ]
+    configurations = [SystemIdentifier(profile, int(line.split()[0]), None) for line in gen_lines]
```
Comment (lukegb) [inline nixos/modules/system/boot/loader/systemd-boot/systemd-boot-builder.py:157]: kwargs please

**Review by lukegb — APPROVED** [review]

File: nixos/modules/system/boot/loader/systemd-boot/systemd-boot-builder.py (hunk 2 above)
Comment (lukegb) [reply nixos/modules/system/boot/loader/systemd-boot/systemd-boot-builder.py:157]:
```suggestion
    configurations = [SystemIdentifier(profile=profile, generation=int(line.split()[0]), specialisation=None) for line in gen_lines]
```

**Comment by sternenseemann** [comment]
@ofborg test systemd-boot

**Review by lukegb — APPROVED** [review]

**Pull request merged**